crossbeam-channel/src/flavors/list.rs

```diff
@@ -538,24 +543,53 @@ impl<T> Channel<T> {
     /// Disconnects senders and wakes up all blocked receivers.
     ///
     /// Returns `true` if this call disconnected the channel.
-    pub(crate) fn disconnect_senders(&self) -> bool {
-        let tail = self.tail.index.fetch_or(MARK_BIT, Ordering::SeqCst);
+    pub(crate) fn disconnect_senders_to_drop(&self) -> bool {
+        let tail = self
+            .tail
+            .index
+            .fetch_or(TAIL_DISCONNECT_IMPLICIT, Ordering::SeqCst);
+
+        if tail & TAIL_DISCONNECT_IMPLICIT == 0 {
+            self.receivers.disconnect();
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn disconnect(&self) -> bool {
+        let tail = self
+            .tail
+            .index
+            .fetch_or(TAIL_DISCONNECT_EXPLICIT, Ordering::SeqCst);
 
-        if tail & MARK_BIT == 0 {
+        if tail & TAIL_DISCONNECT_ANY == 0 {
```

Comment:
the use of TAIL_DISCONNECT_ANY, instead of TAIL_DISCONNECT_EXPLICIT is intentional and needs a comment...